File: src/main/java/aurocosh/divinefavor/common/lib/DistributedRandomList.java
package aurocosh.divinefavor.common.lib;

import java.util.ArrayList;
import java.util.List;

public class DistributedRandomList<T> {
    private DistributedRandom distributedRandom;
    private List<T> values;

    public DistributedRandomList() {
        this.distributedRandom = new DistributedRandom();
        this.values = new ArrayList<>();
    }

    public void add(T value, double distribution) {
        int newIndex = values.size();
        values.add(value);
        distributedRandom.addNumber(newIndex, distribution);
    }

    public void clear() {
        values.clear();
        distributedRandom.clear();
    }

    public T getRandom() {
        if(values.isEmpty())
            return null;
        int index = distributedRandom.getRandomNumber();
        return values.get(index);
    }

    public int size() {
        return values.size();
    }

    public T get(int index) {
        return values.get(index);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/lib/distributed_random/DistributedRandomList.kt
package aurocosh.divinefavor.common.lib.distributed_random

import java.util.*

open class DistributedRandomList<T> {
    private val distributedRandom: DistributedRandom = DistributedRandom()
    private val values: MutableList<T>

    val random: T?
        get() {
            if (values.isEmpty())
                return null
            val index = distributedRandom.randomNumber
            return values[index]
        }

    init {
        this.values = ArrayList()
    }

    fun add(value: T, distribution: Double) {
        val newIndex = values.size
        values.add(value)
        distributedRandom.addNumber(newIndex, distribution)
    }

    fun clear() {
        values.clear()
        distributedRandom.clear()
    }

    fun size(): Int {
        return values.size
    }

    operator fun get(index: Int): T {
        return values[index]
    }
}
